tests-ng/DEVELOPER.md

```diff
@@ -0,0 +1,490 @@
+
+#### `@pytest.mark.performance_metric`
+
+Marks performance tests that can be skipped under emulation:
+
+```python
+@pytest.mark.performance_metric
+@pytest.mark.booted(reason="We can only measure startup time if we actually boot the system")
+def test_startup_time(systemd: Systemd):
+    # Test implementation
+```
+
+### Parametrization
+
+Use `@pytest.mark.parametrize` for testing multiple scenarios:
+
+```python
+@pytest.mark.parametrize("sshd_config_item", required_sshd_config)
+def test_sshd_has_required_config(sshd_config_item: str, sshd: Sshd):
+    actual_value = sshd.get_config_section(sshd_config_item)
+    expected_value = required_sshd_config[sshd_config_item]
+    # Test implementation
+```
+
+### Missing Markers (TODO)
+
+#### `@pytest.mark.security_id`
+
+TODO: Explain what this marker is good for.
+
+## Python Best Practices
+
+### Code Style and CI Enforcement
+
+The project enforces code quality through CI linting (see `.github/workflows/lint_tests-ng.yml`):
+
+- **[Black](https://black.readthedocs.io/en/stable/)**: Automatic code formatting
+- **[isort](https://pycqa.github.io/isort/)**: Import statement sorting
+- **[Pyright](https://github.com/microsoft/pyright/blob/main/docs/getting-started.md)**: Static type checking
+
+**Guidelines:**
+
+- Follow [PEP 8 – Style Guide for Python Code](https://peps.python.org/pep-0008/)
+- Use type hints for function parameters and return values
+- Prefer descriptive variable names
+- Use docstrings for complex functions
+
+> [!TIP]
+> Run `make -f tests-ng/dev.makefile format` locally before committing
+
+### Error Handling
+
+- Use descriptive assertion messages
+- Handle expected failures gracefully
+- Use appropriate exception types
+
+```python
+def test_config_file_exists():
+    config_path = Path("/etc/ssh/sshd_config")
+    assert config_path.exists(), f"SSH config file not found: {config_path}"
+```
+
+### Imports
+
+- Group imports: standard library, third-party, local
+- Use specific imports when possible
+- Avoid wildcard imports
+
+```python
+import difflib
+import gzip
+import json
+import os
+import re
+import shutil
+from dataclasses import asdict, dataclass
+from datetime import datetime
+from pathlib import Path
+from typing import Dict, List, Optional
+
+import pytest
+from debian import deb822
+
+from .dpkg import Dpkg
+from .kernel_module import KernelModule, LoadedKernelModule
+from .shell import ShellRunner
+from .sysctl import Sysctl, SysctlParam
+from .systemd import Systemd, SystemdUnit
+```
+
+## External Dependencies
+
+### Policy
+
+- **Prefer Python standard library** over third-party packages
+- **Only add PyPI dependencies** when there's clear benefit
+- **Document why** external dependencies are necessary
+- **Consider maintenance cost** of external dependencies
+
+### Current Dependencies
+
+See `tests-ng/util/requirements.txt` for current dependencies. Each dependency should be justified.
+
+### Adding New Dependencies
+
+When adding new dependencies:
```

Comment:
With external dependencies we also need to make sure that the versions we are using a safe to use and actively maintained.
In one of my projects, we use a `cargo-audit` CI to help us with that, maybe something similar exists for python and would be helpful with that?

Reply:
https://github.com/gardenlinux/gardenlinux/blob/8d15eb4f9c14870273e43f812758fb97c7646973/.github/dependabot.yml#L20-L26 should already handle updates to our dependencies.

Maybe adding a comment for maintained dependencies is enough, then?

Reply:
Ah, right ... completely forgot about our bot. Yeah should be enough for python. A comment about a dependency receiving regular maintenance should suffice then :)